lib/orders/use_case/update_order_status.rb

```diff
@@ -0,0 +1,22 @@
+require 'orders/domain/order_error'
+
+module Orders
+  module UseCase
+    class UpdateOrderStatus
+      Request = Struct.new(:order_id, :status)
+
+      attr_reader :gateway
+
+      def initialize(gateway: Gateway::Order.new)
+        @gateway = gateway
+      end
+
+      def update_orders_status(request)
+        order = @gateway.get_order(request[:order_id])
+        raise Orders::Domain::OrderError if order.nil?
```

Comment:
I'd avoid raising exceptions in use cases, the cleanest way is to use a presenter pattern. The next best thing is returning in your response an array of error codes.

Reply:
(Unless you expect this to cause a 500)

Reply:
I find this interesting, if in my domain I want to show that this use case failed so that it can be handled appropriately by whoever has called I would have thought an exception would be the cleanest thing to show that.

As an example we may have multiple ports updating order status, one may need to do an action on this error whereas another may ignore it. With this being the case returning an error code would mean investigating it rather than catching the thrown error.

Am I missing something here?

Reply:
Exceptions should be exceptional :D - so if you raise an exception, I'd expect the user to see an internal server error.

Regardless of whether you check the length/contents of an errors array or introduce a try .. catch block, you are "investigating" the error. Except one of these is more dangerous. 

I personally think that the presenter pattern is cleaner as the use case takes an instance of an object that can render various outputs of a use case.

```ruby

presenter = JsonInvoicePresenter.new()
view_invoice.execute({invoice_id: 1}, presenter)

return presenter.to_json()
```

The presenter then has various methods like e.g. `#invoice_total(amount)`, `#line_item(item)`

but it can also have error methods like

`#invoice_not_found()`
`#invalid_vat_treatment()`